streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -160,22 +160,36 @@ class ZSink[-R, -InErr, -In, +OutErr, +L, +Z](val channel: ZChannel[R, InErr, Ch
    *
    * This function essentially runs sinks in sequence.
    */
-  def flatMap[R1 <: R, InErr1 <: InErr, OutErr1 >: OutErr, In1 <: In, L1 >: L, Z1](
+  def flatMap[R1 <: R, InErr1 <: InErr, OutErr1 >: OutErr, In1 <: In, L1 >: L <: In1, Z1](
     f: Z => ZSink[R1, InErr1, In1, OutErr1, L1, Z1]
   )(implicit ev: L <:< In1): ZSink[R1, InErr1, In1, OutErr1, L1, Z1] =
     foldM(ZSink.fail(_), f)
 
-  def foldM[R1 <: R, InErr1 <: InErr, OutErr2, In1 <: In, L1 >: L, Z1](
+  def foldM[R1 <: R, InErr1 <: InErr, OutErr2, In1 <: In, L1 >: L <: In1, Z1](
     failure: OutErr => ZSink[R1, InErr1, In1, OutErr2, L1, Z1],
     success: Z => ZSink[R1, InErr1, In1, OutErr2, L1, Z1]
   )(implicit ev: L <:< In1): ZSink[R1, InErr1, In1, OutErr2, L1, Z1] =
     new ZSink(
       channel.doneCollect.foldM(
         failure(_).channel,
         { case (leftovers, z) =>
-          (ZChannel.write(leftovers.flatMap(_.map(ev))) *> ZChannel.identity[InErr1, Chunk[In1], Any]) >>> success(
-            z
-          ).channel
+          ZChannel.fromEffect(Ref.make(leftovers.flatten)).flatMap { leftoversRef =>
+            val refReader = ZChannel.fromEffect(leftoversRef.getAndSet(Chunk.empty)).flatMap { chunk =>
```

Comment:
It looks like we keep needing Refs for storing leftovers. This is unfortunate. Maybe we should accommodate this in the encoding?

Reply:
It looks like you found a way around this???